File: src-wails/llm_api.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

type TestChatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type TestChatRequest struct {
	Provider     string            `json:"provider"`
	Messages     []TestChatMessage `json:"messages"`
	SystemPrompt string            `json:"systemPrompt"`
	WorldBook    []WorldBookEntry  `json:"worldBook"`
}

type TestChatResponse struct {
	Content string `json:"content"`
	Error   string `json:"error,omitempty"`
}

func (a *App) SendTestChat(request TestChatRequest) TestChatResponse {
	systemPrompt := request.SystemPrompt
	if len(request.WorldBook) > 0 {
		matchedEntries := a.matchWorldBookEntries(request.WorldBook, request.Messages)
		if matchedEntries != "" {
			systemPrompt = systemPrompt + "\n\n" + matchedEntries
		}
	}
	var content string
	var err error
	switch request.Provider {
	case "claude":
		content, err = a.callClaudeAPI(request.Messages, systemPrompt)
	case "gemini":
		content, err = a.callGeminiAPI(request.Messages, systemPrompt)
	case "grok":
		content, err = a.callGrokAPI(request.Messages, systemPrompt)
	default:
		return TestChatResponse{Error: "未知的 API 提供商"}
	}
	if err != nil {
		return TestChatResponse{Error: err.Error()}
	}
	return TestChatResponse{Content: content}
}

func (a *App) matchWorldBookEntries(worldBook []WorldBookEntry, messages []TestChatMessage) string {
	var allText strings.Builder
	for _, msg := range messages {
		allText.WriteString(msg.Content)
		allText.WriteString(" ")
	}
	searchText := strings.ToLower(allText.String())
	var matched []string
	for _, entry := range worldBook {
		if entry.IsFolder {
			continue
		}
		keywords := entry.Keywords
		if len(keywords) == 0 && entry.Key != "" {
			keywords = strings.Split(entry.Key, ",")
		}
		for _, kw := range keywords {
			kw = strings.TrimSpace(strings.ToLower(kw))
			if kw != "" && strings.Contains(searchText, kw) {
				if entry.Value != "" {
					matched = append(matched, entry.Value)
				}
				break
			}
		}
	}
	return strings.Join(matched, "\n\n")
}

func (a *App) callClaudeAPI(messages []TestChatMessage, systemPrompt string) (string, error) {
	config, err := a.LoadConfig()
	if err != nil || config.ClaudeApiKey == "" {
		return "", fmt.Errorf("Claude API Key 未配置")
	}
	claudeMessages := make([]map[string]string, len(messages))
	for i, msg := range messages {
		claudeMessages[i] = map[string]string{
			"role":    msg.Role,
			"content": msg.Content,
		}
	}
	requestBody := map[string]any{
		"model":      "claude-sonnet-4-20250514",
		"max_tokens": 4096,
		"messages":   claudeMessages,
	}
	if systemPrompt != "" {
		requestBody["system"] = systemPrompt
	}
	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return "", err
	}
	req, err := http.NewRequest("POST", "https://api.anthropic.com/v1/messages", bytes.NewBuffer(jsonData))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", config.ClaudeApiKey)
	req.Header.Set("anthropic-version", "2023-06-01")
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	if resp.StatusCode != 200 {
		return "", fmt.Errorf("Claude API 错误: %s", string(body))
	}
	var result struct {
		Content []struct {
			Text string `json:"text"`
		} `json:"content"`
		Error *struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", err
	}
	if result.Error != nil {
		return "", fmt.Errorf(result.Error.Message)
	}
	if len(result.Content) > 0 {
		return result.Content[0].Text, nil
	}
	return "", fmt.Errorf("Claude 返回空响应")
}

func (a *App) callGeminiAPI(messages []TestChatMessage, systemPrompt string) (string, error) {
	config, err := a.LoadConfig()
	if err != nil || config.GeminiApiKey == "" {
		return "", fmt.Errorf("Gemini API Key 未配置")
	}
	var contents []map[string]any
	for _, msg := range messages {
		role := msg.Role
		if role == "assistant" {
			role = "model"
		}
		contents = append(contents, map[string]any{
			"role": role,
			"parts": []map[string]string{
				{"text": msg.Content},
			},
		})
	}
	requestBody := map[string]any{
		"contents": contents,
	}
	if systemPrompt != "" {
		requestBody["systemInstruction"] = map[string]any{
			"parts": []map[string]string{
				{"text": systemPrompt},
			},
		}
	}
	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return "", err
	}
	url := fmt.Sprintf("https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=%s", config.GeminiApiKey)
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	if resp.StatusCode != 200 {
		return "", fmt.Errorf("Gemini API 错误: %s", string(body))
	}
	var result struct {
		Candidates []struct {
			Content struct {
				Parts []struct {
					Text string `json:"text"`
				} `json:"parts"`
			} `json:"content"`
		} `json:"candidates"`
		Error *struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", err
	}
	if result.Error != nil {
		return "", fmt.Errorf(result.Error.Message)
	}
	if len(result.Candidates) > 0 && len(result.Candidates[0].Content.Parts) > 0 {
		return result.Candidates[0].Content.Parts[0].Text, nil
	}
	return "", fmt.Errorf("Gemini 返回空响应")
}

func (a *App) callGrokAPI(messages []TestChatMessage, systemPrompt string) (string, error) {
	config, err := a.LoadConfig()
	if err != nil || config.GrokApiKey == "" {
		return "", fmt.Errorf("Grok API Key 未配置")
	}
	var grokMessages []map[string]string
	if systemPrompt != "" {
		grokMessages = append(grokMessages, map[string]string{
			"role":    "system",
			"content": systemPrompt,
		})
	}
	for _, msg := range messages {
		grokMessages = append(grokMessages, map[string]string{
			"role":    msg.Role,
			"content": msg.Content,
		})
	}
	requestBody := map[string]any{
		"model":    "grok-3-latest",
		"messages": grokMessages,
	}
	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return "", err
	}
	req, err := http.NewRequest("POST", "https://api.x.ai/v1/chat/completions", bytes.NewBuffer(jsonData))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+config.GrokApiKey)
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	if resp.StatusCode != 200 {
		return "", fmt.Errorf("Grok API 错误: %s", string(body))
	}
	var result struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
		Error *struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", err
	}
	if result.Error != nil {
		return "", fmt.Errorf(result.Error.Message)
	}
	if len(result.Choices) > 0 {
		return result.Choices[0].Message.Content, nil
	}
	return "", fmt.Errorf("Grok 返回空响应")
}

File: src-tauri/src/services/llm_api.rs
use crate::types::{ChatRequest, ChatResponse};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize)]
struct ClaudeRequest {
    model: String,
    max_tokens: i32,
    messages: Vec<ClaudeMessage>,
    #[serde(skip_serializing_if = "Option::is_none")]
    system: Option<String>,
}

#[derive(Debug, Serialize)]
struct ClaudeMessage {
    role: String,
    content: String,
}

#[derive(Debug, Deserialize)]
struct ClaudeResponse {
    content: Vec<ClaudeContent>,
}

#[derive(Debug, Deserialize)]
struct ClaudeContent {
    text: String,
}

#[derive(Debug, Serialize)]
struct GeminiRequest {
    contents: Vec<GeminiContent>,
    #[serde(skip_serializing_if = "Option::is_none")]
    system_instruction: Option<GeminiSystemInstruction>,
}

#[derive(Debug, Serialize)]
struct GeminiContent {
    role: String,
    parts: Vec<GeminiPart>,
}

#[derive(Debug, Serialize, Deserialize)]
struct GeminiPart {
    text: String,
}

#[derive(Debug, Serialize)]
struct GeminiSystemInstruction {
    parts: Vec<GeminiPart>,
}

#[derive(Debug, Deserialize)]
struct GeminiResponse {
    candidates: Vec<GeminiCandidate>,
}

#[derive(Debug, Deserialize)]
struct GeminiCandidate {
    content: GeminiCandidateContent,
}

#[derive(Debug, Deserialize)]
struct GeminiCandidateContent {
    parts: Vec<GeminiPart>,
}

pub async fn send_chat(request: ChatRequest) -> Result<ChatResponse, String> {
    let model = request.model.to_lowercase();
    if model.contains("claude") {
        call_claude_api(request).await
    } else if model.contains("gemini") {
        call_gemini_api(request).await
    } else if model.contains("grok") {
        call_grok_api(request).await
    } else {
        Err(format!("Unsupported model: {}", request.model))
    }
}

async fn call_claude_api(request: ChatRequest) -> Result<ChatResponse, String> {
    let api_key = crate::database::db_get_config("claude_api_key")
        .map_err(|e| e.to_string())?
        .unwrap_or_default();
    if api_key.is_empty() {
        return Err("Claude API key not configured".to_string());
    }
    let messages: Vec<ClaudeMessage> = request
        .messages
        .iter()
        .map(|m| ClaudeMessage {
            role: m.role.clone(),
            content: m.content.clone(),
        })
        .collect();
    let claude_request = ClaudeRequest {
        model: request.model,
        max_tokens: request.max_tokens.unwrap_or(4096),
        messages,
        system: request.system_prompt,
    };
    let client = reqwest::Client::new();
    let resp = client
        .post("https://api.anthropic.com/v1/messages")
        .header("x-api-key", &api_key)
        .header("anthropic-version", "2023-06-01")
        .header("Content-Type", "application/json")
        .json(&claude_request)
        .send()
        .await
        .map_err(|e| e.to_string())?;
    if !resp.status().is_success() {
        let error_text = resp.text().await.unwrap_or_default();
        return Ok(ChatResponse {
            content: String::new(),
            error: Some(error_text),
        });
    }
    let claude_response: ClaudeResponse = resp.json().await.map_err(|e| e.to_string())?;
    let content = claude_response
        .content
        .first()
        .map(|c| c.text.clone())
        .unwrap_or_default();
    Ok(ChatResponse {
        content,
        error: None,
    })
}

async fn call_gemini_api(request: ChatRequest) -> Result<ChatResponse, String> {
    let api_key = crate::database::db_get_config("gemini_api_key")
        .map_err(|e| e.to_string())?
        .unwrap_or_default();
    if api_key.is_empty() {
        return Err("Gemini API key not configured".to_string());
    }
    let contents: Vec<GeminiContent> = request
        .messages
        .iter()
        .map(|m| GeminiContent {
            role: if m.role == "assistant" { "model".to_string() } else { "user".to_string() },
            parts: vec![GeminiPart { text: m.content.clone() }],
        })
        .collect();
    let system_instruction = request.system_prompt.map(|s| GeminiSystemInstruction {
        parts: vec![GeminiPart { text: s }],
    });
    let gemini_request = GeminiRequest {
        contents,
        system_instruction,
    };
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/{}:generateContent?key={}",
        request.model, api_key
    );
    let client = reqwest::Client::new();
    let resp = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&gemini_request)
        .send()
        .await
        .map_err(|e| e.to_string())?;
    if !resp.status().is_success() {
        let error_text = resp.text().await.unwrap_or_default();
        return Ok(ChatResponse {
            content: String::new(),
            error: Some(error_text),
        });
    }
    let gemini_response: GeminiResponse = resp.json().await.map_err(|e| e.to_string())?;
    let content = gemini_response
        .candidates
        .first()
        .and_then(|c| c.content.parts.first())
        .map(|p| p.text.clone())
        .unwrap_or_default();
    Ok(ChatResponse {
        content,
        error: None,
    })
}

async fn call_grok_api(request: ChatRequest) -> Result<ChatResponse, String> {
    let api_key = crate::database::db_get_config("grok_api_key")
        .map_err(|e| e.to_string())?
        .unwrap_or_default();
    if api_key.is_empty() {
        return Err("Grok API key not configured".to_string());
    }
    let mut messages: Vec<serde_json::Value> = vec![];
    if let Some(system) = &request.system_prompt {
        messages.push(serde_json::json!({
            "role": "system",
            "content": system
        }));
    }
    for m in &request.messages {
        messages.push(serde_json::json!({
            "role": m.role,
            "content": m.content
        }));
    }
    let grok_request = serde_json::json!({
        "model": request.model,
        "messages": messages,
        "max_tokens": request.max_tokens.unwrap_or(4096)
    });
    let client = reqwest::Client::new();
    let resp = client
        .post("https://api.x.ai/v1/chat/completions")
        .header("Authorization", format!("Bearer {}", api_key))
        .header("Content-Type", "application/json")
        .json(&grok_request)
        .send()
        .await
        .map_err(|e| e.to_string())?;
    if !resp.status().is_success() {
        let error_text = resp.text().await.unwrap_or_default();
        return Ok(ChatResponse {
            content: String::new(),
            error: Some(error_text),
        });
    }
    let grok_response: serde_json::Value = resp.json().await.map_err(|e| e.to_string())?;
    let content = grok_response["choices"][0]["message"]["content"]
        .as_str()
        .unwrap_or_default()
        .to_string();
    Ok(ChatResponse {
        content,
        error: None,
    })
}
